module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,741 @@
+
+	startupCheck bool
+	fetchTimeout time.Duration
+
+	bootstrapped chan struct{}
+}
+
+// NewexecutionDataRequester creates a new execution data requester component
+func New(
+	log zerolog.Logger,
+	edrMetrics module.ExecutionDataRequesterMetrics,
+	datastore datastore.Batching,
+	blobservice network.BlobService,
+	eds state_synchronization.ExecutionDataService,
+	startBlockHeight uint64,
+	maxCachedEntries uint64,
+	maxSearchAhead uint64,
+	blocks storage.Blocks,
+	results storage.ExecutionResults,
+	fetchTimeout time.Duration,
+	startupCheck bool,
+) (ExecutionDataRequester, error) {
+	e := &executionDataRequesterImpl{
+		log:              log.With().Str("component", "execution_data_requester").Logger(),
+		ds:               datastore,
+		bs:               blobservice,
+		eds:              eds,
+		metrics:          edrMetrics,
+		startBlockHeight: startBlockHeight,
+		blocks:           blocks,
+		results:          results,
+		startupCheck:     startupCheck,
+		fetchTimeout:     fetchTimeout,
+		status: status.New(
+			datastore,
+			log.With().Str("component", "requester_status").Logger(),
+			startBlockHeight,
+			maxCachedEntries,
+			maxSearchAhead,
+		),
+
+		finalizedBlocks:    make(chan flow.Identifier, 1),
+		fetchRequests:      make(chan *status.BlockEntry, fetchQueueLength),
+		fetchRetryRequests: make(chan *status.BlockEntry, fetchQueueLength),
+		notifications:      make(chan struct{}, fetchQueueLength),
+		bootstrapped:       make(chan struct{}),
+	}
+
+	builder := component.NewComponentManagerBuilder().
+		AddWorker(e.finalizedBlockProcessor).
+		AddWorker(e.notificationProcessor)
+
+	for i := 0; i < fetchWorkers; i++ {
+		builder.AddWorker(e.fetchRequestProcessor)
+	}
+
+	e.cm = builder.Build()
+	e.Component = e.cm
+
+	return e, nil
+}
+
+// AddOnExecutionDataFetchedConsumer adds a callback to be called when a new ExecutionData is received
+// Callback Implementations must:
+//   * be concurrency safe
+//   * be non-blocking
+//   * handle repetition of the same events (with some processing overhead).
+func (e *executionDataRequesterImpl) AddOnExecutionDataFetchedConsumer(fn ExecutionDataReceivedCallback) {
+	e.consumerMu.Lock()
+	defer e.consumerMu.Unlock()
+	e.consumers = append(e.consumers, fn)
+}
+
+func (e *executionDataRequesterImpl) OnBlockFinalized(block *model.Block) {
+	logger := e.log.With().Str("finalized_block_id", block.BlockID.String()).Logger()
+
+	// stop accepting new blocks if the component is shutting down
+	if util.CheckClosed(e.cm.ShutdownSignal()) {
+		logger.Warn().Msg("ignoring finalized block. component is shutting down")
+		return
+	}
+
+	// limit how far ahead of the notifications we're allowed to get
+	if e.status.IsPaused() {
+		logger.Debug().Msg("ignoring finalized block. requester is paused")
+		return
+	}
+
+	logger.Debug().Msg("received finalized block notification")
+
+	select {
+	case e.finalizedBlocks <- block.BlockID:
+	default:
+		logger.Warn().Msg("finalized block queue is full")
```

Comment:
⚠️  Do we need to guarantee we never miss any block to fetch its execution data?

If so then this will break the guarantee, because if some finalized block is dropped, then we won't fetch for these dropped finalized blocks again, right? 

We ran into the same issue for verification node, where it needs to go through each finalized block, and check (see verificaiton/assigner/block_consumer) if there is any chunk assigned and guarantee it doesn't miss any finalized block, if there chunk assigned, it (see verification/fetcher/chunkconsumer) runs a pool of workers to process each chunk. 

To guarantee it never miss a block, the assigner will trace what is the last finalized height that has been processed, and fetch next unprocessed height when there is worker available. 

There are a few patterns are quite similar, we actually [extracted some modules ](https://github.com/onflow/flow-go/blob/master/module/jobqueue/consumer.go)to be reusable in case it could be applied else where.

Reply:
this will break on the last blocks for spork

Reply:
this code was refactored in b4ce6a5 to use jobqueues to track the work flow